### ISSUE TITLE
workflow not working for batch image processing

### ISSUE THREAD

**Issue by Carl-Linnaeus** [danforthcenter/plantcv#769]
I am trying to batch process images using the plantcv-workflow.py. I have made a workflow using the naive bayes multiclassifier and I am trying to run that in the workflow.py with the workflow.config. I have checked for vailidation of my config and it comes out true. But when I run !python plantcv-workflow.py --config my_test.json in Jupyter with a test folder of 2 images, It reads the image metadata and says the joblist will include 4 images. I tested my naive bayes multiclassifer code and it worked for a single image but not when I try to use the workflow parallelization. 

config.input_dir = "./test_batch"
config.json = "output.json"
config.filename_metadata = ["camera","id", "other"]
config.workflow = "./NB_workflow.py"
config.img_outdir= "./output_images"
config.delimiter ="_"
config.imgformat = "JPG"
config.writeimg = True
config.append = False
config.cluster = "LocalCluster"

I run this line:
!python plantcv-workflow.py --config my_test.json

it starts reading the image metadata but nothing happens for processing images, and I have let it run for over an hour before. So I stop the kernal and this is what I get for the traceback.

Starting run 2021-05-07_16-12-42

Reading image metadata...
Reading image metadata took 0.006010293960571289 seconds.
Building job list... 
Job list will include 4 images
Building job list took 0.00413203239440918 seconds.
Processing images... 
^C
Traceback (most recent call last):
  File "plantcv-workflow.py", line 219, in <module>
    main()
  File "plantcv-workflow.py", line 195, in main
    cluster_client = plantcv.parallel.create_dask_cluster(cluster=config.cluster, cluster_config=config.cluster_config)
  File "/workspace/kohela/plantcv/plantcv/parallel/multiprocess.py", line 31, in create_dask_cluster
    client = Client(n_workers=cluster_config.get("n_workers"))
  File "/home/kohela/.local/lib/python3.7/site-packages/distributed/client.py", line 746, in __init__
    self.start(timeout=timeout)
  File "/home/kohela/.local/lib/python3.7/site-packages/distributed/client.py", line 954, in start
    sync(self.loop, self._start, **kwargs)
  File "/home/kohela/.local/lib/python3.7/site-packages/distributed/utils.py", line 348, in sync
    e.wait(10)
  File "/tools/eb/software/releases/2018b/Python/3.7.0/lib/python3.7/threading.py", line 552, in wait
    signaled = self._cond.wait(timeout)
  File "/tools/eb/software/releases/2018b/Python/3.7.0/lib/python3.7/threading.py", line 300, in wait
    gotit = waiter.acquire(True, timeout)
KeyboardInterrupt

**Expected behavior**
I expected the workflow to run through all the images and give me the information I have entered in my workflow.

**Local environment (please complete the following information):**
 - OS:  Windows
 - Environment : Jupyter Notebooks
 - PlantCV Version: 3.11

Any help on how to fix this is greatly appreciated!

**Comment by nfahlgren** [comment]
Hi @Carl-Linnaeus,

Could you post what a directory listing of `./test_batch` looks like?

Thanks!

**Comment by Carl-Linnaeus** [comment]
Sure! its

['.ipynb_checkpoints', '.JPG', 'IMG_8783_2021-03-09.JPG', 'IMG_8748_2021-03-09.JPG']

thanks!

**Comment by nfahlgren** [comment]
Could you try changing:

```python
config.filename_metadata = ["camera","id", "other"]
```

To

```python
config.filename_metadata = ["camera","id", "timestamp"]
```

**Comment by Carl-Linnaeus** [comment]
Just tried it.

its picking up the ipynb_checkpoints.

Starting run 2021-05-17_20-07-03

Reading image metadata...
unconverted data remains: -checkpoint
  --> Please specify the correct timestampformat parameter <--

**Comment by nfahlgren** [comment]
Oh actually I forgot something else. You would also need to set:

```python
config.timestampformat = "%Y-%m-%d"
```

**Comment by Carl-Linnaeus** [comment]
I had changed that when I ran it.
here was the updated code:
config.input_dir = "./test_batch"
config.json = "output.json"
config.filename_metadata = ["camera","id","timestamp"]
config.workflow = "./NB_workflow.py"
config.img_outdir= "./output_images"
config.delimiter ="_"
config.timestampformat="%Y-%m-%d"
config.timestamp="%Y-%m-%d"
config.imgformat = "jpg"
config.writeimg = True
config.append = False

Why would the folder "test_batch" have the checkpoints and JPG? I am thinking that may be the reason it is not working.
['.ipynb_checkpoints', '.JPG', 'IMG_8783_2021-03-09.JPG', 'IMG_8748_2021-03-09.JPG']

Any ideas?

Thanks!

**Comment by nfahlgren** [comment]
The `.ipynb_checkpoints` file is created by Jupyter when a notebook is run from that location. I am not sure about the file`.JPG`. These can probably be deleted.

I think `imgformat` is case-sensitive so should be "JPG"

I just want to double check that after updating `config` in the notebook you save it to the file `my_test.json` using:

```python
config.save_config("my_test.json")
```

**Comment by Carl-Linnaeus** [comment]
yes I run that right before I try to run the workflow.

**Comment by nfahlgren** [comment]
Could you try setting:

```python
config.tmp_dir = "./tmp"
config.cleanup = False
```

And then checking what the contents of `./tmp` look like? It should contain a JSON file for each input image with the image metadata (and later PlantCV results) in the contents.

**Comment by Carl-Linnaeus** [comment]
I tried that but nothing is in those files. Probably because I cannot remove the '.ipynb_checkpoints' and '.JPG' from the  "./test_batch" directory. I have tried os.remove and os.removedirs but without any success.

**Comment by nfahlgren** [comment]
What are the names of the files in `./tmp`?

**Comment by Carl-Linnaeus** [comment]
plantcv/tmp/2021-05-17_22-00-15__oqy04t6
plantcv/tmp/2021-05-17_22-00-46_dv80mrky

I tried running it a couple times

**Comment by nfahlgren** [comment]
Oh right, it creates temp directories per run. What's in one of those?

**Comment by Carl-Linnaeus** [comment]
there is nothing in either

**Comment by nfahlgren** [comment]
Would you mind creating a zip archive of your workflow and two example images and either uploading it here or emailing it to me? I need to test what is causing the issue and may need to inject some debugging steps in

**Comment by JorgeGtz** [comment]
Hi @Carl-Linnaeus,
It seems that one problem comes from parsing what is inside '.ipynb_checkpoints'.
Could you try setting:
`config.include_all_subdirs = False`

**Comment by Carl-Linnaeus** [comment]
Hi  @JorgeGtz,

That did work to reduce the number of images to the actual number of 2, instead of the 4 that I was getting. Thanks!

Starting run 2021-05-18_17-24-48

Reading image metadata...
Reading image metadata took 0.00542759895324707 seconds.
Building job list... 
Job list will include 2 images
Building job list took 0.0025970935821533203 seconds.
Processing images... 

But the workflow still did not run. 


Hi @nfahlgren, 

With what @JorgeGtz suggested I did see the json file show up in the tmp folder. Shown below:

{"metadata": {"camera": {"label": "camera identifier", "datatype": "<class 'str'>", "value": "IMG"}, "imgtype": {"label": "image type", "datatype": "<class 'str'>", "value": "none"}, "zoom": {"label": "camera zoom setting", "datatype": "<class 'str'>", "value": "none"}, "exposure": {"label": "camera exposure setting", "datatype": "<class 'str'>", "value": "none"}, "gain": {"label": "camera gain setting", "datatype": "<class 'str'>", "value": "none"}, "frame": {"label": "image series frame identifier", "datatype": "<class 'str'>", "value": "none"}, "lifter": {"label": "imaging platform height setting", "datatype": "<class 'str'>", "value": "none"}, "timestamp": {"label": "datetime of image", "datatype": "<class 'datetime.datetime'>", "value": "2021-03-09"}, "id": {"label": "image identifier", "datatype": "<class 'str'>", "value": "8783"}, "plantbarcode": {"label": "plant barcode identifier", "datatype": "<class 'str'>", "value": "none"}, "treatment": {"label": "treatment identifier", "datatype": "<class 'str'>", "value": "none"}, "cartag": {"label": "plant carrier identifier", "datatype": "<class 'str'>", "value": "none"}, "measurementlabel": {"label": "experiment identifier", "datatype": "<class 'str'>", "value": "none"}, "other": {"label": "other identifier", "datatype": "<class 'str'>", "value": "none"}, "image": {"label": "image file", "datatype": "<class 'str'>", "value": "/kohela/plantcv/test_batch/IMG_8783_2021-03-09.JPG"}}, "observations": {}}

But it did not show any of the information that I wanted from the workflow (i.e. the output images and pixels colors) I can send you a zip file to your email if that works for you. Thanks!

**Comment by nfahlgren** [comment]
Hi @Carl-Linnaeus, could you upload your workflow script here? We think the issue is that it's stalling somehow (email works too though)

**Comment by Carl-Linnaeus** [comment]
`
# Naive bayes workbook for batch image analysis

import os
import cv2
import numpy as np
from scipy import stats
from matplotlib import pyplot as plt
from plantcv import plantcv as pcv
import argparse

def options():
    """Parse command-line options
    The options function was converted from the class options in Jupyter.
    Rather than hardcoding the inputs, input arguments with the same
    variable names are used to retrieve inputs from the commandline.
    """
    parser = argparse.ArgumentParser(description="NB batch workflow")
    parser.add_argument("-i","--image", help="Input image", required=True)
    parser.add_argument("-r","--result", help="Results file", required=True)
    parser.add_argument("-o","--outdir", help="Output directory", required=True)
    parser.add_argument("-w","--writeimg", help="Save output images", action="store_true")
    parser.add_argument("-D","--debug", help="Plot", default=None)
    args = parser.parse_args()
    return args

def main():
    
    args = options()
    pcv.params.debug = args.debug
    pcv.params.debug_outdir = args.outdir
   

   
    img, path, filename = pcv.readimage(filename=args.image)
    img = pcv.rotate(img, 90, False)

    mask = pcv.naive_bayes_classifier(rgb_img=img, 
                                      pdf_file="Biomass_test_pdfs.txt")

    green_img = pcv.apply_mask(mask=(mask['Green']), img=img, mask_color='black')
    yellow_img = pcv.apply_mask(mask=(mask['Yellow']), img=img, mask_color='black')
    background_img = pcv.apply_mask(mask=(mask['Background']), img=img, mask_color='black')
    stake_img = pcv.apply_mask(mask=(mask['Stake']), img=img, mask_color='black')


    green_maskpath, plant_analysis_images = pcv.output_mask(img=img, mask=mask['Green'], 
                                                            filename='green.png', mask_only=True)
    yellow_maskpath, pust_analysis_images = pcv.output_mask(img=img, mask=mask['Yellow'], 
                                                          filename='yellow.png', mask_only=True)
    background_maskpath, chlor_analysis_images = pcv.output_mask(img=img, mask=mask['Background'], 
                                                            filename='background.png', mask_only=True)
    stake_maskpath, bkgrd_analysis_images = pcv.output_mask(img=img, mask=mask['Stake'], 
                                                            filename='stake.png', mask_only=True)


    classified_img = pcv.visualize.colorize_masks(masks=[mask['Green'], mask['Yellow'], 
                                                         mask['Background'], mask['Stake']], 
                                                  colors=['dark green', 'yellow', 'black', 'gray'])

    # Calculate percent of the plant biomass including yellow leaves
    biomass_plant = np.count_nonzero(mask['Green']) + np.count_nonzero(mask['Yellow'])
    non_plant = np.count_nonzero(mask['Background']) + np.count_nonzero(mask['Stake'])
    percent_biomass = biomass_plant/(biomass_plant+non_plant)
    percent_biomass

    total_pixels=biomass_plant+non_plant
    total_pixels

    #calculate cm2 
    #100 cm =  3709 pixels
    #1375.668 pixels =cm2
    cmsqrd=1375.668
    totalcmsqrd=biomass_plant/cmsqrd
    totalcmsqrd

    pcv.outputs.add_observation(sample='default', variable='cm2', trait='area (cm2) of plant in image',
                                method='biomass_plant/cmsqrd', scale='area', datatype=float,
                                value=totalcmsqrd, label='cm2')

    # Data stored to the outputs class can be accessed using the variable name
    pcv.outputs.observations['default']['cm2']['value']
    
    pcv.plot_image(classified_img)

    #pcv.print_results(filename=args.result)
    
    #seeing if this code works for parallelization
    
    pcv.print_results(filename = args.result+"_"+str(i)+".txt")
    with open(args.result, "a") as finalfile:
            with open('prefix_' +str(i)+ '.txt') as tempfile:
                for x in tempfile.readlines():
                    finalfile.write(x)
                    finalfile.write("\n")
                tempfile.close()
            os.remove('prefix_' +str(i)+ '.txt')
    finalfile.close()
 

# re-enble line below for debugging
    pcv.plot_image(classified_img)                   
    pcv.plot_image(img_copy)
    
if __name__=='__main__':
    main()`

**Comment by nfahlgren** [comment]
At the end you should be able to replace:

```python
#pcv.print_results(filename=args.result)

#seeing if this code works for parallelization

pcv.print_results(filename = args.result+"_"+str(i)+".txt")
with open(args.result, "a") as finalfile:
        with open('prefix_' +str(i)+ '.txt') as tempfile:
            for x in tempfile.readlines():
                finalfile.write(x)
                finalfile.write("\n")
            tempfile.close()
        os.remove('prefix_' +str(i)+ '.txt')
finalfile.close()
```

with:

```python
pcv.outputs.save_results(args.result)
```

Note: `pcv.print_results()` still works but will be deprecated in favor of `pcv.outputs.save_results()` in a later version

**Comment by Carl-Linnaeus** [comment]
I tried it a couple times but I am still getting the Processing images... Should it take more than 10 minutes to clock time?

**Comment by nfahlgren** [comment]
It depends on the workflow. But if you ran the workflow in Jupyter and it took less than 10 minutes then it shouldn't take that long in parallel. I got your email but wasn't able to extract the images for some reason, but was able to with the workflow.

One other thing to try on your end. If there is a directory called dask_workspace or similar could you remove it and try running plantcv-workflow.py again?

**Comment by Carl-Linnaeus** [comment]
in the plantcv/plantcv/parallel/multiprocess.py? remove it from the .py then try running it again? I see a create_dask_cluster in that code.

**Comment by nfahlgren** [comment]
Oh nope, I meant dask creates a temporary working directory. Sometimes I have had weird behavior when that directory still exists after a previous run

**Comment by Carl-Linnaeus** [comment]
I had only tested it as a notebook before and that worked. But after going through many of the plantCV tutorials again I saw the sample command to run a workflow on a single image:

`./workflowname.py -i testimg.png -o ./output-images -r results.txt -w -D 'print'`

So I tested that out with my workflow and it did not work. 

`!python /workspace/kohela/plantcv/NB_workflow.py -i /workspace/kohela/plantcv/test_batch/IMG_8748_2021-3-09.JPG -o /workspace/kohela/plantcv/output-images -r results.txt -w -D 'plot'`

The reults I am getting:

`Premature end of JPEG file
Figure(640x480)
Figure(640x480)
Figure(640x480)
Figure(640x480)
Figure(640x480)
Figure(640x480)
Figure(640x480)
Figure(640x480)
Figure(640x480)
Figure(640x480)
Figure(640x480)
Figure(640x480)
Figure(640x480)
Figure(640x480)
Figure(640x480)
Figure(640x480)`

No images are produced and the -r results.txt never shows up.
the main problem I think is the "Premature end of JPEG file" any thoughts on this? Does it need to be a .png file for the workflow to be used or is it the filename? Maybe the date in the filename?

Thank you!
